File: Leetcode Daily Challenge/027-September-2025/006-minimum-operations-to-make-array-elements-zero.cpp
// https://leetcode.com/problems/minimum-operations-to-make-array-elements-zero/




#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    using ll = long long;

    ll getSteps(ll l, ll r) {
        // L...R   : Steps = ?  // R = L * 4 - 1
        // 1...3   : Steps = 1
        // 4...15  : Steps = 2
        // 16...63 : Steps = 3
        
        ll L = 1, step = 1;
        ll stepsToReachZero = 0;

        while (r >= L) {
            ll R = L * 4 - 1;

            ll start = max(L, l);
            ll end   = min(R, r);

            if (start <= end) 
                stepsToReachZero += ((end - start + 1) * step);

            step += 1;
            L    *= 4;
        }

        return stepsToReachZero;
    }
public:
    // TC : O(Q * log4(max(qr)))
    // SC : O(1)
    long long minOperations(vector<vector<int>>& queries) {
        ll result = 0;

        for (vector<int> &query : queries) {
            ll l = query[0], r = query[1];

            ll steps = getSteps(l, r);

            result += (steps + 1) / 2;  // ceil value
        }

        return result;
    }
};

File: Leetcode Daily Challenge/027-September-2025/006-minimum-operations-to-make-array-elements-zero.java
// https://leetcode.com/problems/minimum-operations-to-make-array-elements-zero/




class Solution {
    // TC : O(Q * log4(max(qr)))
    // SC : O(1)
    public long minOperations(int[][] queries) {
        long result = 0;
        for (int[] query : queries) {
            long l = query[0], r = query[1];

            long steps = getSteps(l, r);

            result += (steps + 1) / 2;
        }
        return result;
    }

    private static long getSteps(long l, long r) {
        long L = 1, steps = 1, stepsToReachZero = 0;

        while (r >= L) {
            long R = L * 4 - 1;

            long start = Math.max(L, l);
            long end   = Math.min(R, r);

            if (start <= end) {
                stepsToReachZero += ((end - start + 1) * steps);
            }

            L     *= 4;
            steps += 1;
        }

        return stepsToReachZero;
    }
}